source/transfer_list.rst

```diff
@@ -657,5 +658,65 @@ such that the last ACPI table in this entry ends at offset
      - hdr_size
      - One or more ACPI tables.
 
+
+.. _tpm_evlog_entry:
+
+TPM info table entry layout (XFERLIST_TPM_INFO)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This entry type holds TPM-related information for a platform. The TPM event log
+info is a region containing a TPM event log as defined by TCG EFI Protocol
+Specification (https://trustedcomputinggroup.org/resource/tcg-efi-protocol-specification/).
+The CRB info defines the address of a region of memory that has been carved out
+and reserved for use as a TPM Command Response Buffer interface.
+
+.. _tab_tpm_info:
+.. list-table:: TPM info type layout
+   :widths: 2 2 2 8
+
+   * - Field
+     - Size (bytes)
+     - Offset (bytes)
+     - Description
+
+   * - tag_id
+     - 0x3
+     - 0x0
+     - The tag_id field must be set to **5**.
+
+   * - hdr_size
+     - 0x1
+     - 0x3
+     - |hdr_size_desc|
+
+   * - data_size
+     - 0x4
+     - 0x4
+     - 0x19
+
+   * - evlog_base_address
+     - 0x8
+     - 0x8
+     - The physical address of the region that contains the TPM event log.
+
+   * - crb_base_address
+     - 0x8
+     - 0x10
+     - The base address of CRB.
```

Comment:
We should not be putting addresses in there...instead the contents of that region should be included in the TE

Reply:
This field needs to contain the physical address of the CRB.

Suggested re-wording:

> The physical base address of a region of memory reserved for use as a TPM's Command Response Buffer region.

As described in the overview text, this is a region of memory that has been carved out and reserved, for example by Trusted Firmware-A.  So, there is no "contents" that would make sense here.  What we want is a way to pass the address of the carved out region between firmware components.

Reply:
Then shouldn't this region be set up within the Transfer List by TF-A? Which program is intended to add this entry to the TL?

Reply:
Yes, exactly.  TF-A would allocate the reserved region and then add the address/size to the transfer list.  This communicates the info to EDK2/u-boot/secure-partitions so they know where the CRB region is.

Reply:
So then, how about:

1. TF-A creates a TL
2. TF-A adds the TPM block to it (say it is 4KB long)
3. TF-A passed it on to U-Boot, etc.?

i.e. why does it need to be in another area of memory?

Reply:
This reserved CRB region will be reserved for use by the TPM implementation over the life of the system, and so it's likely you don't want it coming out of a random Normal world address.

Having it in a separate area gives platform's much more flexibility in when and how they reserve memory.  Some platforms may want to statically define reserved regions at specific addresses (for example: https://github.com/ARM-software/arm-trusted-firmware/blob/master/include/plat/arm/common/arm_def.h).  Others may want to reserve regions in a linker script.  There may be other platform specific approaches.

So, we don't want to force all platforms into one approach, unless there is some compelling reason to do this.

1. TF-A platform code reserves CRB memory by platform specific means
2. TF-A adds the address/size to the TL in the TPM info entry
3. TF-A passes it on to U-boot, etc.

Reply:
The TL is intended to hold the actual data so that should be the default approach.

If you put a secure-world address in a normal-world data structure, what happens? Who could actually consume it?

Reservation of the bloblist memory needs to be handled by common code, with other code adding TEs as needed. So the TF-A platform-specific means can add a TE, rather than adding it in a random address.

At least that should be the default.

Reply:
I sent a PR as I now see that the spec didn't include this language, although it was much-discussed and is documented in the U-Boot bloblist implementation.

Reply:
There are some use-cases where allowing pointers may make sense.

I see 2 compeling ones:
1) some data-structure that can be appended to during boot (e.g. the TPM log). The entity that appends is unaware of the TL, or we don't want to handle potential resizes.
2) Some particular MMIO region could be at a particular fixed address (for some platforms/SoCs the CRB can be hardcoded to some address).

I agree that, in general, TL entries should contain the data.
For the TPM log (and TPM CRB) I wonder if we should open an exception -- as it may be infeasible (for some platforms) to have them within the TL entry.

Reply:
> If you put a secure-world address in a normal-world data structure, what happens? Who could actually consume it?

A secure world address can not be accessed by normal world, so there would never be a reason to pass an address like that to the normal world.

Reply:
Re the use cases:
1. We can allocate the TPM log to be large enough for initial use.  It is possible to expand a TE.
2. You want to put the TPM log in in a memory-mapped I/O device? What type of device would that be?

I am OK with an exception where needed...but then I think we should have two TEs, the normal case (containing the actual TPM log) and the exception one (for when that is not possible and has to be in strange memory). The reason for needing an exception should not be, for example, "our project has platform-specific code which we don't control and wants to put it whereever it likes".

Reply:
One distinction we should make is differentiating between passing "data" and a reference to a "resource".

In the case of the proposed TPM Info TE there are 2 separate items that are currently specified by address/size:
1.  TPM event log.  This is "data" that is populated by TF-A measured boot and passed to next boot stage.
2. CRB address/size.   This is a "resource".  There is no data in it.  It's a reference to a reserved region allocated in a platform specific way and we need to pass the address/size to the next stage of boot.  I can't think of any scenario where it would make sense to allocate this in the TL itself.  It's just an empty buffer.

Right now the TPM info combines both items in one TPM info TE.  I would be ok splitting this into 2 distinct TEs, one for the TPM event log and one for the CRB address/size.

Reply:
OK, so perhaps we can agree that the event log should be embedded in a TE, so we should split into two TEs.

For the CRB, that seems OK...but why can't each program use their own CRB address/size? What is special about that address/size? E.g. I might want to use a local var as the buffer, or allocate it with malloc() ? Sorry I don't know much about CRB , but am happy to read a page or two if you can post a link.

Reply:
There is normally 1 TPM in a system and the CRB is the interface that software will use to send commands to the TPM and receive responses.  So, for Linux the address of the CRB is advertised in an ACPI table.  So software (e.g. Linux) can't pick where the CRB is, it is advertised via ACPI along with other info about the TPM.

If the CRB is in normal DRAM, the most convenient thing to do will be for TF-A to reserve/carve-out a chunk of memory early and then communicate the address/size down to other firmware components.

Not sure there is a good high level summary.
- Linux CRB driver is here: https://github.com/torvalds/linux/blob/master/drivers/char/tpm/tpm_crb.c
- TCG Mobile CRB spec is here: https://trustedcomputinggroup.org/resource/tpm-2-0-mobile-command-response-buffer-interface-specification/.  But, it's more about what the fields in the CRB are and less about how it integrates into a system and who allocates the memory.  There is an intro section that may help.

Reply:
OK, thanks...so are you saying that if the TPM were on PCI we would just use the BARs to find the CRB, but because we are not using PCI, we need to use ACPI (or presumably DT on non-server ARM devices) to specify the address?

It seems from the docs there are two cases:
- TPM chip with MMIO at a particular address
- software TPM with a mailbox at a particular address

Is the address set up using some hardware registers somewhere, like a aperture that you need to select? Can you point to the ACPI and DT bindings?

Reply:
> OK, thanks...so are you saying that if the TPM were on PCI we would just use the BARs to find the CRB, but because we are not using PCI, we need to use ACPI (or presumably DT on non-server ARM devices) to specify the address?
Yes, the address needs to be advertised by ACPI/DT.

>  It seems from the docs there are two cases:
> TPM chip with MMIO at a particular address
> software TPM with a mailbox at a particular address

For a TPM chip accessed through hardware registers, ACPI/DT will describe it like any other device.  It is the "software TPM" case where a DRAM memory carve out is needed for the CRB.

> Is the address set up using some hardware registers somewhere, like a aperture that you need to select?
No.  For a TPM chip on SPI or I2C, it is addressed like any other SPI/I2C device.  There is no setting of the address.  And the hardware addressing info has to be reflected in ACPI/DT.

For the case of a software TPM in TrustZone, the CRB is just a shared memory communication buffer.  The software TPM learns about the address through a mechanism like the transfer list.  TPM clients like u-boot/EDK2/Linux would learn about the location of CRB through the transfer list, ACPI, or DT.

The ACPI binding is here: https://trustedcomputinggroup.org/resource/tcg-acpi-specification/
I am not aware of a DTB binding for TPM CRB yet.  But, we will need one in the case of a TrustZone-based TPM with a client that uses DT.

Reply:
OK, so I believe you need a TE for SW_TPM_CRB which just has the address and size of this region.

For the buffer, there should be a separate TE the actually contains the buffer.

Reply:
This is updated in the new patch

Reply:
Do I have to click on "Resolve Conversation" one i have made the required change?

Reply:
I don't know, honestly. Github reviewing is just bizarre to me.

Someone above suggested "The physical base address of a region of memory reserved for use as a TPM's Command Response Buffer region."

but it doesn't look like you added that?